### PR TITLE
[RISCV][TTI] Implement instruction cost for vp.reduce.* 

### PR DESCRIPTION
**Pull request by ElvisWang123**
This patch based on #114180.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -1179,6 +1179,79 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
     return getCmpSelInstrCost(Instruction::Select, ICA.getReturnType(),
                               ICA.getArgTypes()[0], CmpInst::BAD_ICMP_PREDICATE,
                               CostKind);
+  case Intrinsic::vp_reduce_add:
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1182]: Please see CachingVPExpander::expandPredicationInReduction as an example of how to perform these mappings more cleanly.

Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1182]: Updated, thanks!

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1182]: Please use:

```
Intrinsic::ID RedID = *VPI.getFunctionalIntrinsicID();
unsigned Opc = getArithmeticReductionInstruction(RedID);
OR
getMinMaxReductionIntrinsicOp(RedID);
```

Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1194]: Fixed, thanks!!

File: llvm/test/Analysis/CostModel/RISCV/reduce-add.ll [hunk 2]
```diff
@@ -1,8 +1,10 @@
 ; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
-; RUN: opt < %s -mtriple=riscv32 -mattr=+v -passes="print<cost-model>" -cost-kind=throughput 2>&1 -disable-output | FileCheck %s
-; RUN: opt < %s -mtriple=riscv64 -mattr=+v -passes="print<cost-model>" -cost-kind=throughput 2>&1 -disable-output | FileCheck %s
-; RUN: opt < %s -mtriple=riscv32 -mattr=+v -passes="print<cost-model>" -cost-kind=code-size 2>&1 -disable-output | FileCheck %s --check-prefix=SIZE
-; RUN: opt < %s -mtriple=riscv64 -mattr=+v -passes="print<cost-model>" -cost-kind=code-size 2>&1 -disable-output | FileCheck %s --check-prefix=SIZE
+; RUN: opt < %s -mtriple=riscv32 -mattr=+v -passes="print<cost-model>" -cost-kind=throughput 2>&1 -disable-output --type-based-intrinsic-cost=true | FileCheck %s
```
Comment (preames) [inline llvm/test/Analysis/CostModel/RISCV/reduce-add.ll:2]: Please revert the change (in this and other files) to add --type-based-intrinsic-cost=true.

**Review by preames — CHANGES_REQUESTED** [review]
I am not sure the fully duplicated test coverage for vp.reduce is justified here.  The whole point of this change is that they're cost just like normal reduces?  Why do we need to complete double all test coverage?

**Comment by ElvisWang123** [comment]
> I am not sure the fully duplicated test coverage for vp.reduce is justified here. The whole point of this change is that they're cost just like normal reduces? Why do we need to complete double all test coverage?

The test cases here want to make sure the instruction costs for vp intrinsics are same as their non-vp counterpart.

Adding `type-based-intrinsic-cost=true` let the printer class query the cost model without values which align to how loop vectorizer query the instruction costs.

File: llvm/test/Analysis/CostModel/RISCV/reduce-add.ll (hunk 2 above)
Comment (ElvisWang123) [reply llvm/test/Analysis/CostModel/RISCV/reduce-add.ll:2]: You mean remove `--type-based-intrinsic-cost=true` from the original `RUN` and add another version of the `RUN` line with `--type-based-intrinsic-cost=true`?  Or you thought we have better approach to test the instruction cost of vp-intrinsics here?
I am not  sure what you want?

Comment (preames) [reply llvm/test/Analysis/CostModel/RISCV/reduce-add.ll:2]: You modified the run lines, please remove the modification in this change.  If you want to modify them, do that in it's own change.  (I am skeptical of your motivation for that separate change.)

Comment (ElvisWang123) [reply llvm/test/Analysis/CostModel/RISCV/reduce-add.ll:7]: Reverted. 
Add new run lines for the `type-based` queries and check the instruction costs are same as normal query.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -1191,6 +1191,40 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
...
+    std::optional<Intrinsic::ID> RedID =
+        VPIntrinsic::getFunctionalIntrinsicIDForVP(ICA.getID());
+    assert(RedID.has_value());
+    unsigned RedOp = getArithmeticReductionInstruction(*RedID);
+    if (RedOp == Instruction::FAdd || RedOp == Instruction::FMul)
+      return getArithmeticReductionCost(RedOp,
+                                        cast<VectorType>(ICA.getArgTypes()[1]),
+                                        ICA.getFlags(), CostKind);
+    return getArithmeticReductionCost(
+        RedOp, cast<VectorType>(ICA.getArgTypes()[1]), std::nullopt, CostKind);
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1210]: Why can't we pass ICA.getFlags() for the non FP reductions?

Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Fixed.
Before #114180 landed , passing FMF flags into `getArightmeicReductionCost()` with int reductions will always return the cost of `oredered fadd reduction`.

**Comment by ElvisWang123** [comment]
> I've made specific review suggestions you have not applied. As a means to make forward progress here, are you okay with me landing a modified version of this patch? The code is fine, it's only the testing bits we're stuck on. I'll simply make the testing change I've requested, and you can revisit testing improvements in a separate review if you want.

I push a new commit [2c0f6f1](https://github.com/llvm/llvm-project/pull/114184/commits/2c0f6f10a201846ceac19355ad2641269cb192d6) to revert all RUN line changes.

If the latest change still can not fit your requirement, feel free to land a modified version of this patch.

**Comment by LiqinWeng** [comment]
why close it~

**Comment by lukel97** [comment]
> why close it~

It was landed separately in 3431d133ccfa75d16964be455238e2a1da0c2004
